### ISSUE TITLE
Support for sharded datasets (e.g WebDataset API) for efficiently working on network-attached filesystems

### ISSUE THREAD

**Issue by nikvaessen** [speechbrain/speechbrain#580]
First of all: thanks for making such an amazing tool public!

I was interested in using the VoxCeleb speaker recognition recipe as a baseline/starting point for a research project. However, a major issue is how the data is stored on disk. If I understand correctly SpeechBrain follows the pattern that you have a CSV/JSON manifest, wherein each training sample has a path to a (e.g audio) file on a filesystem mounted on the machine.

In the case of VoxCeleb1 and 2, this amounts to ~1.200.000 files in a single folder. On a local NVMe SSD this is fine, but our cluster uses a network-attached filesystem for providing most storage space. This has severe performance implications - loading ~100.000 audio files (a single epoch of VoxCeleb1) with a network-attached filesystem takes an hour. There is a fairly straightforward solution to this problem: storing your data in so-called shards (e.g TFRecords in Tensorflow). In PyTorch there is a current proposal [1] to include the WebDataset [2, 3] library as an API for sharded datasets. The basic premise of this library is storing your data in multiple `.tar` or `.tar.gz` files.  In my case, this reduced the loading time of a single VoxCeleb1 epoch from an hour to 3 minutes. 

I wanted to discuss a potential integration of WebDatasets in SpeechBrain. Currenty,  `speechbrain.dataio.datasets.DynamicItemDataset` inherits from `torch.util.data.Dataset`, while the `WebDataset` class uses the `torch.util.data.IterableDataset` API. I am not yet familiar enough with the code so I had the following question(s):

* Is there an existing solution for efficiently using speechbrain on network-attached file systems?
* If not, would there be interested in merging a PR offering support for `WebDataset`?
* Any suggestions for the best way to integrate `WebDataset`?

links:
[1] https://github.com/pytorch/pytorch/issues/38419
[2] https://github.com/tmbdev/webdataset
[3] https://pytorch.org/blog/efficient-pytorch-io-library-for-large-datasets-many-files-many-gpus/

**Comment by TParcollet** [comment]
@Gastron @popcornell

**Comment by Gastron** [comment]
# Efficient data IO on NFS

First, thank you for such an insightful post! When working on shared network-attached storage systems (let's call them NFS, even if that might be slightly imprecise), storing and reading data really needs to be handled with care; naive solutions which load many small files may not just be slow for the user, but it can also slow the filesystem down for everyone else. As I understand it, the NFS issue is really just with accessing too many files, and single file throughput is still very fast. Then there's additionally the enormous scale applications where data needs to reside in some separate data warehouse - however I would be content with just solving the "many small files" NFS problem, too, and at the enormous scale, there's.

### Where are we now

You got it right, SpeechBrain does not currently provide any solution for efficient NFS data reading. SpeechBrain has DynamicItemDataset, and when it's applicable I personally think it is very flexible. I guess we could say that that flexibility is afforded by being unaware of streams of data, and instead treating every data point separately - essentially, the difference between map-style and iterable-style Datasets. As you said, this is primarily applicable when there's a fast local storage medium available. Many clusters have fast local storage on the compute nodes, and it's fast enough to copy the data over.

Regarding NFS IO: We just have not yet been confident and happy enough about any particular solution. So in the meanwhile, we opted to just remain compatible with PyTorch Data loading - the rest of SpeechBrain, most importantly the Brain class, should already play nice with all the existing PyTorch compatible solutions, like WebDataset.

### What's on the wish list for the NFS data IO solution?
In no particular order:

- At least we must reduce the number of files, so some kind of sharded storage format is needed.
- Storing the audio (or other data) raw, without feature-extraction. So far in SpeechBrain we've felt it's simple and fast to do feature-extraction and augmentation on-the-fly.
    - Storing any kind of data.
- Ability to handle complex and hierarchical formats for tasks such as diarization or source separation. Would like to be able to refer to sections in many audio files and load all of those sections as one data point.
- Shuffling the shards and shuffling within a shard. Similarly, sorting.
- Seamless and smart co-operation with distributed loading and distributed training. This probably means that each data loading process knows to automatically pick a different shard to load from. The number of loading processes is num_workers times number of training processes.
- Bucketing similar length utterances together.
    - Sorting the data can achieve this.
- Saving the current data loading state for intra-epoch checkpoints.
    - What can suffice: ability to skip ahead a number of examples. If data loading is super fast, maybe it's enough to load a number of data points without processing them.
    - Requires reproducible shuffling
- Random access. Realistically this will need to be somehow constrained, but I believe it should in theory be possible to have random access within a shard.
    - One use case for this is shuffling, but shuffling could be handled with some per-epoch operations too.
    - If not random access, some kind of filtered access could be almost as good.
    - Random access to just some metadata could be good as well.
- Compressed storage for the shards. 
    - However, disk space is relatively cheap and data throughput is relatively fast, so the compression should not come at the cost of a clunky interface.
- Ability to share data prep easily. This can mean a straight-forward script that works for the standard dataset download. With the JSON manifest files for DynamicItemDataset, our plan is to have those manifest files downloadable directly, so no dataprep needed.
- Flexible on the fly transformations that are defined in Python. Hopefully the data pipeline from `speechbrain.utils.data_pipeline` could be used directly.

This is off the top of my head, feel free to add to this. We're going to need to find a reasonable compromise, this is wish list.

### Which solutions could be considered

Hopefully nothing fully custom to SpeechBrain is needed. We should also retain compatibility with the standard PyTorch data IO interfaces. Perhaps the most realistic solution is to augment some outside tool with e.g. the ability to save intra-epoch checkpoints and possibly adding a "DynamicItem"-like interface on top and providing data prep scripts for the common datasets.

- [WebDataset](https://github.com/tmbdev/webdataset) looks like a very strong contender at this point, particularly as it is on track to being included in PyTorch. I would love to see examples of it being used with SpeechBrain, and I think I'll try it out myself too.
- [Lhotse](https://lhotse.readthedocs.io/) is a Kaldi-descendant data IO library, with a focus on speech. The flavor might be a bit different to what SpeechBrain does, but I expect it to be very efficiently engineered.
- [Apache Arrow / Huggingface datasets](https://github.com/huggingface/datasets) This is focused on text. In the background it uses an Arrow database. Unclear if suitable for speech at all.
- **Filesystem in a file** such as HDF5 or more simply ZIP like with [DareBlopy](https://github.com/podgorskiy/DareBlopy) or TAR like WebDataset does.

Would be happy to hear other ideas here as well.

**Comment by Gastron** [comment]
Regarding how to go forward with WebDataset right now: as I said, I believe it should already work with SpeechBrain, but it would just lack some of the dataio features that we have. So I again encourage to try it out.

- WebDataset seems to have a dict interface (?), which would then already play nicely together with SpeechBrain's `PaddedBatch`.
- Some kind of solution for intra-epoch checkpoints should be found. This might mean adding a some special functionality in `Brain`, which would tell the `wds.Dataset` to fast-forward to a certain point in the epoch. This would also require reproducible shuffling.
- WebDataset seems to have it's own data transformation interface. If there's any point in adding the "DynamicItem"-style data pipeline on top, that can be done with some simple interface, but we if the WebDataset transformation is already doing the same thing, we should be happy to just use that.
- A big part of the work with libraries like WebDataset is the data prep - and I don't think there is a generic solution, it depends on the dataset and task. In practice, it would be valuable to see some speech examples of WebDataset, as the examples they provide seem to be just images.

**Comment by popcornell** [comment]
I think @Gastron covered up this almost extensively. 

> Some kind of solution for intra-epoch checkpoints should be found. This might mean adding a some special functionality in Brain, which would tell the wds.Dataset to fast-forward to a certain point in the epoch. This would also require reproducible shuffling.

I think this is the only critical thing. Together with sorting and dynamic batch size which can however be handled in the data prep part directly because we can assume this will be used in such circumstances where your dataset is so big you cannot fit a list of the examples in memory ( not even their path and lenghts). 

By looking at it, it should be pretty much easily integrable regarding other functionalities.

**Comment by nikvaessen** [comment]
Just FYI: I'm working on a sharded version (with Webdataset) of the VoxCeleb recipe here: https://github.com/nikvaessen/speechbrain/tree/sharded-voxceleb/my-recipes/SpeakerRec

I'll give an update when it's in a usable state :)

**Comment by Gastron** [comment]
Keeping this open, as we're actively exploring options and are happy to hear more thoughts from the community as well.

**Comment by tmbdev** [comment]
> Some kind of solution for intra-epoch checkpoints should be found. This might mean adding a some special functionality in `Brain`, which would tell the `wds.Dataset` to fast-forward to a certain point in the epoch. This would also require reproducible shuffling.

Just a couple of comments:

- all shuffling in WebDataset is done using the standard random number generator by default, so if you seed that, it is reproducible
- optionally, you can provide a separate random number generator object just for WebDataset, making shuffling completely independent of any other part of the training pipeline and making it easy to pickle the RNG.

As far as restarting the I/O pipeline, obviously, you could simply keep the RNG state and current count-in-epoch, but that would be slow. Fast restarting is possible by just keeping track of the RNG state, the current shard and the index in the current shard in the I/O pipeline.

I'm curious, though: there are other sources of non-determinism in distributed training jobs and I/O; how are you dealing with those?

**Comment by Gastron** [comment]
Hi @tmbdev, thanks for the comments and thank you for webdataset! Regarding sources of non-determinism in distributed jobs: I feel that I may be missing something, would be great to hear if you can point out some of our blind spots.

The determinism I'd like to see in SpeechBrain is that even with restarting, the training proceeds through all the same batches. There can be non-determinism in async floating point operations, and even with the same seed, the result might not be the exact same model. That is fine to me, but what I'd ideally like to be able to report in a paper is "we trained the model for 50 epochs on the training set", and take that to mean that during training, each example was seen exactly 50 times (and not have to worry about restarts making this statement inaccurate). Though if that is not possible (and in many cases it may just not be so relevant), we can instead report something like "we trained the model for 3 million steps on the training set", meaning just 3 million updates.

Having now used webdataset for a couple of weeks, I feel more confident that it is the way to go here. I was particularly happy to notice that webdataset has a similar `dict` example format, so it worked well with the SpeechBrain `PaddedBatch` class right out of the box. The way we could further "support" webdataset in SpeechBrain is to provide some iterator extensions. With variable length sequences, an on-the-fly bucketing iterator could be very useful. That could also be combined with dynamic size batching. Let's see where we go, if these implementations turn out well, we'd be glad to make a PR in webdataset, too.

**Comment by tmbdev** [comment]
Thanks.

WebDataset just delivers training samples sequentially, and it does so exactly throughout its entire pipeline. The main difficulty is that PyTorch's DDP requires an exactly equal number of batches in all workers right now, and that's a limitation you need to work around, no matter what loader you use. This isn't something the loader can fix, since samples, batches, and nodes just don't divide evenly in general.

You can get as close to a solution as indexed datasets by using Tensorcom; with Tensorcom, all the batches are aggregated from all the loaders before being redistributed to the GPUs. But you're still left with a partial set of batches for end of the epoch, just like with indexed datasets.

A full solution necessarily requires some kind of inter-node communications. One idea is to train with zero length batches in all the nodes that have run out of data and have DDP signal when all nodes have run out of data to end the epoch. It would be a nice little project outside WebDataset and would benefit all loaders.

I've never really understood the purpose of going through that trouble though. The only reason for one-sample-per-epoch has been the relationship between full gradient descent and SGD, but training doesn't approximate full gradient descent anyway (not even close). Furthermore, if you drop the one sample per epoch requirement, you're basically just training with samplinig-with-replacement, which is statistically a perfectly good approach (arguably, better than using each sample exactly once). I generally prefer "# samples seen" as a consistent and easy to compare metric in all experiments.

(Note that the two issues, namely determinism and exact epochs, are separate.)

**Comment by Gastron** [comment]
Right, right. Not caring about exact epochs could be the way to go for now; it is much easier to start with that.

**Comment by mravanelli** [comment]
This features is now available in the development branch. A tutorial is coming as well.
